### PR TITLE
[autogluon][build] AutoGluon 0.6.1 inference containers

### PR DESCRIPTION
**Pull request by gradientsky**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [x] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_pre_release.py [hunk 1]
```diff
@@ -480,7 +481,8 @@ def _run_dependency_check_test(image, ec2_connection):
         "autogluon": {
             "0.3": ["cpu", "gpu"],
```
Comment (kenny-ezirim) [inline test/dlc_tests/sanity/test_pre_release.py:482]: Maybe we should remove as well all versions that are no longer supported?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 2]
```diff
@@ -451,7 +451,8 @@ def _run_dependency_check_test(image, ec2_connection):
...
```
Comment (kenny-ezirim) [inline test/dlc_tests/sanity/test_pre_release.py:452]: Do you still want to keep versions that are no longer supported?

File: test/dlc_tests/container_tests/bin/security_checks.py [hunk 3]
```diff
@@ -26,11 +26,14 @@ def check_that_cache_dir_is_removed(home_dir):
     if os.path.exists(cache_dir_path):
         content_of_cache_dir = [f for f in os.listdir(cache_dir_path)]
         LOGGER.info("Contents of cache directory: %s", content_of_cache_dir)
-        if len(content_of_cache_dir) > 1:
-            raise ValueError("cache dir includes more than 1 file (not only motd)")
-        if not content_of_cache_dir[0].startswith("pip"):
-            raise ValueError("cache dir include file that it probably should not have: {}"
-                             .format(content_of_cache_dir[0]))
+        if len(content_of_cache_dir) == 2 and sorted(content_of_cache_dir) == ['mim', 'pip']:
+            pass  # autogluon
```
Comment (kenny-ezirim) [inline test/dlc_tests/container_tests/bin/security_checks.py:30]: Please add more details here, like "pip and mim are both required for AG" and more details about the files if necessary.

File: autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu.os_scan_allowlist.json [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      "cvss_v3_severity": "CRITICAL",
+      "source_url": "https://people.canonical.com/~ubuntu-security/cve/2022/CVE-2022-45907.html",
+      "source": "UBUNTU_CVE",
+      "severity": "MEDIUM",
+      "status": "ACTIVE",
+      "title": "CVE-2022-45907 - torch",
+      "reason_to_ignore": "the patch is not released yet"
+    }
+  ]
+}
```
Comment (xncqr) [inline autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu.os_scan_allowlist.json:62]: Recommend new line on end of file for best practices. note the red circle in the PR on the bottom of this file.

File: autogluon/inference/docker/0.6/py3/Dockerfile.cpu.os_scan_allowlist.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      "cvss_v3_severity": "CRITICAL",
+      "source_url": "https://people.canonical.com/~ubuntu-security/cve/2022/CVE-2022-45907.html",
+      "source": "UBUNTU_CVE",
+      "severity": "MEDIUM",
+      "status": "ACTIVE",
+      "title": "CVE-2022-45907 - torch",
+      "reason_to_ignore": "the patch is not released yet"
+    }
+  ]
+}
```
Comment (xncqr) [inline autogluon/inference/docker/0.6/py3/Dockerfile.cpu.os_scan_allowlist.json:62]: Recommend new line on end of file for best practices. note the red circle in the PR on the bottom of this file.

File: autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ && pip install --no-cache-dir -U certifi \
+ && pip uninstall -y dataclasses \
+ && pip install --no-cache-dir -U numpy numba \
+ && pip install --no-cache-dir -U torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} torchtext==${TORCH_TEXT_VER} --extra-index-url https://download.pytorch.org/whl/cu113 \
+ && pip install --no-cache-dir -U autogluon==${AUTOGLUON_VERSION} \
+ && mim install mmcv-full \
+ && pip install --no-cache-dir -U mmdet
+
+# Downgrade protobuf 3.20.1 -> 3.19.6: https://pyup.io/v/51167/f17
+RUN pip install -U protobuf==3.19.*
```
Comment (liangfu) [inline autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu:37]: This downgrade seems to be unnecessary to v0.6.1, see https://github.com/autogluon/autogluon/pull/2193

Comment (gradientsky) [reply autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu:37]: It is necessary because the existing version range in autogluon is not supporting safe version of 3.20.x (which is 3.20.6, I believe)

File: test/dlc_tests/sanity/test_pre_release.py (hunk 2 above)
Comment (gradientsky) [reply test/dlc_tests/sanity/test_pre_release.py:452]: removed

File: test/dlc_tests/container_tests/bin/security_checks.py (hunk 3 above)
Comment (gradientsky) [reply test/dlc_tests/container_tests/bin/security_checks.py:30]: added

File: autogluon/inference/docker/0.6/py3/Dockerfile.cpu.os_scan_allowlist.json (hunk 5 above)
Comment (gradientsky) [reply autogluon/inference/docker/0.6/py3/Dockerfile.cpu.os_scan_allowlist.json:62]: fixed

File: autogluon/inference/docker/0.6/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ && pip uninstall -y dataclasses \
+ && pip install --no-cache-dir -U certifi \
+ && pip install --no-cache-dir -U numpy numba \
+    torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} torchtext==${TORCH_TEXT_VER} --extra-index-url https://download.pytorch.org/whl/cpu \
+ && pip install --no-cache-dir -U autogluon==${AUTOGLUON_VERSION} \
+ && mim install -q mmcv-full \
+ && pip install --no-cache-dir -U mmdet
+
+# Downgrade protobuf 3.20.1 -> 3.19.6: https://pyup.io/v/51167/f17
+RUN pip install -U protobuf==3.19.*
```
Comment (liangfu) [inline autogluon/inference/docker/0.6/py3/Dockerfile.cpu:44]: Same here.

This downgrade seems to be unnecessary to v0.6.1, since we've upgrade torch==1.11.0. See https://github.com/autogluon/autogluon/pull/2193

Comment (gradientsky) [reply autogluon/inference/docker/0.6/py3/Dockerfile.cpu:44]: ditto - see above

File: autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu.os_scan_allowlist.json (hunk 4 above)
Comment (gradientsky) [reply autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu.os_scan_allowlist.json]: fixed

**Review by kenny-ezirim — APPROVED** [review]

**Review by liangfu — APPROVED** [review]
LGTM! Thanks for the contribution.

File: autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu (hunk 6 above)
Comment (liangfu) [reply autogluon/inference/docker/0.6/py3/cu112/Dockerfile.gpu:37]: Just a quick note per offline discussion, it is the requirement from sagemaker-training 4.3.1 that requires protobuf=3.19.x
```
sagemaker-training 4.3.1 requires protobuf<3.20,>=3.9.2, but you have protobuf 4.21.12 which is incompatible.
```

File: test/dlc_tests/sanity/test_pre_release.py [hunk 8]
```diff
@@ -229,7 +229,7 @@ def test_framework_version_cpu(image):
         if tested_framework == "autogluon.core":
             versions_map = {
                 # container version -> autogluon version
-                '0.3.2': '0.3.1',
+                # '0.3.2': '0.3.1',
```
Comment (arjkesh) [inline test/dlc_tests/sanity/test_pre_release.py:232]: does this need to be removed?

File: test/dlc_tests/container_tests/bin/security_checks.py [hunk 9]
```diff
@@ -26,11 +26,16 @@ def check_that_cache_dir_is_removed(home_dir):
     if os.path.exists(cache_dir_path):
         content_of_cache_dir = [f for f in os.listdir(cache_dir_path)]
         LOGGER.info("Contents of cache directory: %s", content_of_cache_dir)
-        if len(content_of_cache_dir) > 1:
-            raise ValueError("cache dir includes more than 1 file (not only motd)")
-        if not content_of_cache_dir[0].startswith("pip"):
-            raise ValueError("cache dir include file that it probably should not have: {}"
-                             .format(content_of_cache_dir[0]))
+        if len(content_of_cache_dir) == 2 and sorted(content_of_cache_dir) == ['mim', 'pip']:
```
Comment (arjkesh) [inline test/dlc_tests/container_tests/bin/security_checks.py:29]: will this allow these for all containers, not just AG? is there a way to make an AG specific exception so that we do not add this rule everywhere?

File: autogluon/buildspec.yml [hunk 10]
```diff
@@ -1,8 +1,8 @@
 account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
Comment (arjkesh) [inline autogluon/buildspec.yml:1]: please remove inference images from this file

File: test/test_utils/__init__.py [hunk 11]
```diff
@@ -1671,9 +1671,13 @@ def construct_buildspec_path(dlc_path, framework_path, buildspec, framework_vers
         else:
             raise ValueError(f"Framework version {framework_version} was not matched.")
 
-    buildspec_path = os.path.join(dlc_path, framework_path, f"{buildspec}.yml")
+    # for backward compatibility, first try new path structure 
+    # if it fails then revert to prior structure
+    buildspec_path = os.path.join(dlc_path, framework_path, job_type, f"{buildspec}.yml")
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1676]: NoneType was relevant in initial implementation. I think it should be empty string instead with current implementation.

File: test/test_utils/__init__.py [hunk 12]
```diff
@@ -1642,7 +1642,7 @@ def get_python_version_from_image_uri(image_uri):
     return "py36" if python_version == "py3" else python_version
 
 
-def construct_buildspec_path(dlc_path, framework_path, buildspec, framework_version):
+def construct_buildspec_path(dlc_path, framework_path, buildspec, framework_version, job_type = None):
```
[the remark this answers is not in the document]
Comment (kenny-ezirim) [reply test/test_utils/__init__.py:1645]: Agreed.

File: test/dlc_tests/sanity/test_pre_release.py (hunk 8 above)
Comment (kenny-ezirim) [reply test/dlc_tests/sanity/test_pre_release.py:232]: Yes, they are no longer supported.

Comment (kenny-ezirim) [reply test/dlc_tests/sanity/test_pre_release.py:232]: @gradientsky Can you confirm this?

File: test/dlc_tests/container_tests/bin/security_checks.py (hunk 9 above)
Comment (kenny-ezirim) [reply test/dlc_tests/container_tests/bin/security_checks.py:29]: @gradientsky comment?

File: test/dlc_tests/sanity/test_pre_release.py (hunk 8 above)
Comment (gradientsky) [reply test/dlc_tests/sanity/test_pre_release.py:232]: I left this in place as an example: just in case if this functionality would be required in future.

File: test/dlc_tests/container_tests/bin/security_checks.py (hunk 9 above)
Comment (gradientsky) [reply test/dlc_tests/container_tests/bin/security_checks.py:29]: Current codebase doesn't have an easy way to differentiate the logic based on framework.

**Comment by gradientsky** [comment]
Rebased from master

**Review by arjkesh — APPROVED** [review]
